src/index.ts

```diff
@@ -23,108 +39,203 @@ export {
 	}
 
+	extract.on('error', (error: Error) => {
+		console.log('resin-bundle-resolve extract on error', error);
+		pack.emit('error', error);
+	});
+
 	extract.on(
 		'entry',
 		async (header: tar.Headers, stream: Readable, next: () => void) => {
-			const name = Utils.normalizeTarEntry(header.name);
+			try {
+				console.log('resin-bundle-resolve extract on entry');
 
-			// Handle the case where the dockerfile is specified
-			if (dockerfile != null) {
-				if (specifiedFileResolver == null && name === dockerfile) {
-					specifiedFileResolver = await resolveSpecifiedFile(
-						resolvers,
+				specifiedFileResolver =
+					(await resolveTarStreamOnEntry(
+						{ header, stream, next },
 						bundle,
-						name,
-						header,
-						stream,
+						resolvers,
 						pack,
-					);
-					next();
-					return;
-				}
-			} else {
-				const potentials = resolvers.filter(r => r.needsEntry(name));
-				if (potentials.length > 0) {
-					const entry = await streamToFileInfo(stream, header);
-					for (const resolver of potentials) {
-						resolver.entry(entry);
-					}
-					// Also add it to the stream
-					pack.entry(header, entry.contents);
-					next();
-					return;
-				}
+						specifiedFileResolver,
+						dockerfile,
+					)) || specifiedFileResolver;
+			} catch (error) {
+				console.log('resin-bundle-resolve extract entry -> error', error);
+				pack.emit('error', error);
 			}
-			stream.pipe(pack.entry(header));
-			stream.on('end', next);
 		},
 	);
 
-	extract.on('finish', async () => {
-		// Ensure that at least one resolver is satisfied, otherwise emit an error
-
-		// Firstly, check that if the user specified a dockerfile, and that dockerfile has
-		// been found and processed
-		let resolver: Resolver;
-
-		if (dockerfile != null) {
-			if (specifiedFileResolver == null) {
-				pack.emit(
-					'error',
-					new Error(
-						`Specified dockerfile could not be resolved: ${dockerfile}`,
-					),
-				);
-				return;
-			}
-			resolver = specifiedFileResolver;
-		} else {
-			// Detect if any of the resolvers have been satisfied
-			const satisfied = _(resolvers)
-				.filter(r => r.isSatisfied(bundle))
-				.orderBy('priority', 'desc')
-				.value();
-
-			if (satisfied.length === 0) {
-				pack.emit('error', new Error('Resolution could not be performed'));
-				return;
-			}
+	extract.once('finish', async () => {
+		try {
+			console.log('resin-bundle-resolve extract on finish');
 
-			resolver = satisfied[0];
-			try {
-				await addResolverOutput(bundle, resolver, pack);
-			} catch (e) {
-				pack.emit('error', e);
-				return;
+			await resolveTarStreamOnFinish(
+				bundle,
+				resolvers,
+				pack,
+				specifiedFileResolver,
+				dockerfile,
+			);
+		} catch (error) {
+			console.log('resin-bundle-resolve extract finish -> error', error);
+			pack.emit(error);
+		} finally {
+			console.log(
+				'resin-bundle-resolve extract on finish calling pack.finalize()',
+			);
+			pack.finalize();
+		}
```

Comment:
Some errors thrown in stream listeners were not being caught and `pack.finalize()` not called, so I added the try-catch-finally above and "forwarded" the errors through `pack.emit('error')`.